### ISSUE TITLE
Datadog ITR doesn't play well with context hooks

### ISSUE THREAD

**Issue by lucas-telus** [DataDog/datadog-ci-rb#256]
**Current behaviour**

We're running a large RSpec suite with tens of thousands of examples, and to improve performance and developer experience, we use several patterns like the ones described in [test-prof/test-prof](https://github.com/test-prof/test-prof), many of which use context hooks under the hood (like `before_all` and `let_it_be`).

While both Datadog's CI visibility and test optimization features provide excellent observability, we've noticed that for many scenarios the performance improvements aren't as significant we expected. This is because RSpec is running context hooks (preparing test setups) even for test files that will be skipped anyway.

We're reporting this as a bug because we found the behavior of running hooks for skipped tests unexpected.

**Expected behaviour**

Ideally, before running a context, Datadog CI should check if all tests within that context will be skipped. If so, it should not execute context callbacks.

Some relevant RSpec internals, just for reference:
 - https://github.com/rspec/rspec-core/blob/v3.13.2/lib/rspec/core/example_group.rb#L603

**Steps to reproduce**

Here's a minimal example to illustrate the issue:

```rb
RSpec.describe do
  before(:all) do # context hook, scoped to this specific test file (or rather block)
    # this will run only once, before executing the tests inside the block
    sleep 5
  end

  it do
    puts "Skipped by datadog ITR"
  end
end
```

**Environment**

* **datadog-ci-rb version:** 1.7.0
* **Configuration block (`Datadog.configure ...`):** -
* **Test framework version:** RSpec v3.13.0
* **CI Provider:** -
* **Ruby version:** -
* **Operating system:** -

**Comment by lucas-telus** [comment]
Thanks for clarifying this!

> it is an implementation detail of RSpec instrumentation: we skip by test, not by test suite, so we have to run the test suite (otherwise you won't see the skipped tests in the Datadog UI at all)

I see that ITR uses the `skip` metadata to skip examples. It's also possible to set the `skip` metadata at the example group level to skip all examples within that group. The behavior is identical to adding skip to individual examples, plus it automatically skips context level hooks. Datadog UI seems to handle this approach fine. Do you see any immediate concerns with this?

**Comment by lucas-telus** [comment]
For example:

```rb
RSpec.describe 'Foo', skip: 'Determined that Datadog ITR would skip all descendant examples - skipping early' do
  before(:all) do 
    sleep 5
  end

  # Example skip message overrides group-level skip message in RSpec reports and Datadog UI
  it 'is skipped', skip: 'Skipped by Datadog ITR' do 
    puts 'Skipped by Datadog ITR'
  end
end
```

**Comment by anmarchenko** [comment]
Yes, setting skip metadata on example group level is exactly the best solution to this issue and it will improve ITR a lot, should be a no-brainer to do. Thanks for bringing this issue up, the only reason it isn't like that already is just because I didn't think about this 😓 

Give me some time to add a PR for this, hopefully next week I'll have something if a bus doesn't hit me.

**Comment by anmarchenko** [comment]
@lucas-telus 1.10.0 is released with this fix - please tell me if it works for you!

**Comment by lucas-telus** [comment]
Thanks for tackling it so quickly @anmarchenko! Looking forward to testing the new version, will let you know how it goes.
